From d300e7f66f9a42fc9fa0700a9960c37f6f537a52 Mon Sep 17 00:00:00 2001
From: Steve Dickson <SteveD@redhat.com>
Date: Thu, 26 Jun 2014 15:54:16 -0400
Subject: [net] sunrpc/xprtrdma: Fall back to MTHCAFMR when FRMR is not supported

Message-id: <1403798077-28799-58-git-send-email-steved@redhat.com>
Patchwork-id: 84926
O-Subject: [RHEL7.1 PATCH 57/78] xprtrdma: Fall back to MTHCAFMR when FRMR is not supported
Bugzilla: 1113248
RH-Acked-by: J. Bruce Fields <bfields@redhat.com>

From: Chuck Lever <chuck.lever@oracle.com>

An audit of in-kernel RDMA providers that do not support the FRMR
memory registration shows that several of them support MTHCAFMR.
Prefer MTHCAFMR when FRMR is not supported.

If MTHCAFMR is not supported, only then choose ALLPHYSICAL.

Signed-off-by: Chuck Lever <chuck.lever@oracle.com>
Signed-off-by: Anna Schumaker <Anna.Schumaker@Netapp.com>
Upstream Id: f10eafd3a6ce9da7e96999c124b643ea6c4921f3
Bugzilla: 1113248
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/net/sunrpc/xprtrdma/verbs.c b/net/sunrpc/xprtrdma/verbs.c
index 6bb9a07..a352798 100644
--- a/net/sunrpc/xprtrdma/verbs.c
+++ b/net/sunrpc/xprtrdma/verbs.c
@@ -491,33 +491,32 @@ rpcrdma_ia_open(struct rpcrdma_xprt *xprt, struct sockaddr *addr, int memreg)
   ia->ri_dma_lkey = ia->ri_id->device->local_dma_lkey;
  }
 
- switch (memreg) {
- case RPCRDMA_MTHCAFMR:
-  if (!ia->ri_id->device->alloc_fmr) {
-   dprintk("RPC:       %s: MTHCAFMR registration "
-    "specified but not supported by adapter, "
-    "using riskier RPCRDMA_ALLPHYSICAL\n",
-    __func__);
-   memreg = RPCRDMA_ALLPHYSICAL;
-  }
-  break;
- case RPCRDMA_FRMR:
+ if (memreg == RPCRDMA_FRMR) {
   /* Requires both frmr reg and local dma lkey */
   if ((devattr.device_cap_flags &
        (IB_DEVICE_MEM_MGT_EXTENSIONS|IB_DEVICE_LOCAL_DMA_LKEY)) !=
       (IB_DEVICE_MEM_MGT_EXTENSIONS|IB_DEVICE_LOCAL_DMA_LKEY)) {
    dprintk("RPC:       %s: FRMR registration "
-    "specified but not supported by adapter, "
-    "using riskier RPCRDMA_ALLPHYSICAL\n",
-    __func__);
-   memreg = RPCRDMA_ALLPHYSICAL;
+    "not supported by HCA\n", __func__);
+   memreg = RPCRDMA_MTHCAFMR;
   } else {
    /* Mind the ia limit on FRMR page list depth */
    ia->ri_max_frmr_depth = min_t(unsigned int,
     RPCRDMA_MAX_DATA_SEGS,
     devattr.max_fast_reg_page_list_len);
   }
-  break;
+ }
+ if (memreg == RPCRDMA_MTHCAFMR) {
+  if (!ia->ri_id->device->alloc_fmr) {
+   dprintk("RPC:       %s: MTHCAFMR registration "
+    "not supported by HCA\n", __func__);
+#if RPCRDMA_PERSISTENT_REGISTRATION
+   memreg = RPCRDMA_ALLPHYSICAL;
+#else
+   rc = -EINVAL;
+   goto out2;
+#endif
+  }
  }
 
  /*
-- 
1.7.1